### PR TITLE
[FLINK-20081][connector/common][source] Fix the executor notifier to let the handler run in main thread when handling exception from the callable.

### PR DESCRIPTION
**Pull request by becketqin**
## What is the purpose of the change
This patch fixes the following two issues:
1. Currently the `ExecutorNotifier` runs the `handler` in the worker thread if there is an exception thrown from the `callable`. This breaks the threading model and prevents an exception from bubbling up to fail the job.
2. When an exception bubbles up from the `SourceCoordinator`, the `UncaughtExceptionHandler` will call `System.exit(-17)` and kill the JM. This is too much. Instead, we should just fail the job to trigger a failover.

## Brief change log
- Let the the `handler` always run in the main thread in `ExecutorNotifier`
- Fail the job instead of killing the JM in the `UncaughtExceptionHandler` in the `SourceCoordinator` executor.

## Verifying this change
The following unit tests have been added to verify the change.
- ExecutorNotifierTest.testExceptionInHandlerWhenHandlingException()
- SourceCoordinatorProviderTest.testCallAsyncExceptionFailsJob()

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by becketqin** [comment]
@lirui-apache @StephanEwen The patch is created to address the issue we discussed over slack when implementing the Hive source. Would you have time to take a look? Thanks.

**Review by lirui-apache — APPROVED** [review]
LGTM. Thanks @becketqin !

**Comment by lirui-apache** [comment]
One minor suggestion is perhaps we should clarify error handling in the `SplitEnumeratorContext::callAsync` java doc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java [hunk 1]
```diff
@@ -83,8 +83,8 @@ public ExecutorNotifier(ScheduledExecutorService workerExecutor,
 				T result = callable.call();
 				executorToNotify.execute(() -> handler.accept(result, null));
 			} catch (Throwable t) {
-				LOG.error("Unexpected exception {}", t);
-				handler.accept(null, t);
+				LOG.error("Unexpected exception", t);
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java:86]: Remove / update this log line.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java [hunk 2]
```diff
@@ -133,7 +133,8 @@ public ExecutorNotifier(ScheduledExecutorService workerExecutor,
 				T result = callable.call();
 				executorToNotify.execute(() -> handler.accept(result, null));
 			} catch (Throwable t) {
-				handler.accept(null, t);
+				LOG.error("Unexpected exception", t);
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/ExecutorNotifier.java:136]: Remove / update the log line.

**Comment by StephanEwen** [comment]
The fix is good.

I would suggest to remove the indicated log statements. This is an exception happening in a utility function. The function lacks the context whether the exception really is a problem or not. Maybe the connector is just trying to do something and it is fine if it fails. But with the log statement there, we always have ERROR level messages coming out of the system on the master node, which looks super scary to users and folks running Flink.

In general, I would avoid logging in utility calls.
Logging that does not add information (the infamous "Unexpected Exception" log messages) are usually a good indication that this log statement should be removed. It "double reports" the exception, once to the log (without any context of whether it means something or not) and to the handler (which has the context whether the exception means something or not).

**Comment by becketqin** [comment]
@lirui-apache @StephanEwen Thanks for the review. I have addressed all the comments. Will merge the patch after CI passes.

**Pull request closed without merging**